File: leetcode/75_LeetCode_Question/Array/2.Best_Time_to_Buy_and_Sell_Stock/2.Best_Time_to_Buy_and_Sell_Stock.py
#!/bin/python3
#https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
from typing import List

'''given an array of prices of a stock, the indices of each prices are months
Calculate the maximum profit when buy and sell the stock at different months'''

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        min_prices = 100000
        for price in prices:
            if price < min_prices: min_prices = price
            elif (price - min_prices) > max_profit:
                max_profit = price- min_prices
        return max_profit

if __name__ == "__main__":
    test4 = Solution()
    print(test4.maxProfit([7,6,4,3,1]))

File: leetcode/75_LeetCode_Question/Array/2.Best_Time_to_Buy_and_Sell_Stock/2.Best_Time_to_Buy_and_Sell_Stock.cpp
//!/usr/bin/gdb
//https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
#include<stdio.h>
#include<math.h>

int maxProfit( int* prices,int pricesSize){
    int max_profit, min_price=prices[0];
    for(int i=1; i<pricesSize; i++)
    {
        int price = prices[i];
        if (min_price>price) min_price = price;
        if ((price-min_price) > max_profit)
        {
            max_profit = price-min_price;
        }
    }
    return max_profit;
}

int main(){
    int prices[5] = {7,6,4,3,1};
    printf("%d\n",maxProfit(prices,5));
    return 0;
}
